File: 2022/day23/day23.exs
#!/usr/bin/env elixir
# Copyright 2022 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.
#
# https://adventofcode.com/2022/day/23

defmodule Day23 do
  @moduledoc """
  Input is a grid of elf (`#`) and space (`.`) positions.  Each round the elves
  stay put if they aren't adjacent to any elf, then move in a cardinal direction
  if there are no elves in the three cells in that direction.  Elves don't move
  if they would collide.  The direction preference cycles each round.
  """

  @north {-1, 0}
  @south {1, 0}
  @east {0, 1}
  @west {0, -1}
  @ne {-1, 1}
  @nw {-1, -1}
  @se {1, 1}
  @sw {1, -1}
  @stay {0, 0}
  @northern {@north, [@north, @ne, @nw]}
  @southern {@south, [@south, @se, @sw]}
  @western {@west, [@west, @nw, @sw]}
  @eastern {@east, [@east, @ne, @se]}
  @stay_put {@stay, [@stay]}
  @all_dirs [@nw, @north, @ne, @east, @se, @south, @sw, @west]

  @doc "Count the empty spaces after 10 rounds."
  def part1(input) do
    points = parse_input(input, 1, MapSet.new())
    pref_cycle = Stream.cycle([@northern, @southern, @western, @eastern])

    points =
      Enum.reduce(1..10, points, fn round, points ->
        run_round(points, round_prefs(round, pref_cycle))
      end)

    if Enum.count(points) < 10, do: print_points(points)
    bounding_rectangle_size(points) - Enum.count(points)
  end

  @doc "Identify the first round when the elves have all stopped moving."
  def part2(input) do
    points = parse_input(input, 1, MapSet.new())
    pref_cycle = Stream.cycle([@northern, @southern, @western, @eastern])

    Enum.reduce_while(Stream.iterate(1, &(&1 + 1)), points, fn round, points ->
      next = run_round(points, round_prefs(round, pref_cycle))
      if MapSet.equal?(points, next), do: {:halt, round}, else: {:cont, next}
    end)
  end

  defp run_round(points, prefs) do
    Enum.reduce(points, %{}, fn point, acc ->
      dir = pick_move(point, points, prefs)
      Map.update(acc, move(point, dir), [point], fn rest -> [point | rest] end)
    end)
    |> Enum.map(fn
      {dest, [_cur]} -> [dest]
      {_, several} -> several
    end)
    |> List.flatten()
    |> Enum.into(MapSet.new())
  end

  defp round_prefs(round, pref_cycle),
    do: Stream.drop(pref_cycle, rem(round - 1, 4)) |> Stream.take(4) |> Enum.into([])

  defp bounding_rectangle(points) do
    {top, bottom} = points |> Enum.map(&elem(&1, 0)) |> Enum.min_max()
    {left, right} = points |> Enum.map(&elem(&1, 1)) |> Enum.min_max()
    {top, bottom, left, right}
  end

  defp bounding_rectangle_size(points) do
    {top, bottom, left, right} = bounding_rectangle(points)
    (bottom - top + 1) * (right - left + 1)
  end

  defp pick_move(point, points, prefs) do
    if Enum.all?(@all_dirs, fn dir -> empty?(move(point, dir), points) end) do
      @stay
    else
      {dir, _} =
        Enum.find(prefs, @stay_put, fn {_, dirs} ->
          dirs |> Enum.map(&move(point, &1)) |> Enum.all?(&empty?(&1, points))
        end)

      dir
    end
  end

  defp empty?(point, points), do: not MapSet.member?(points, point)

  defp move({row, col}, {drow, dcol}), do: {row + drow, col + dcol}

  defp parse_input([], _row, acc), do: acc

  defp parse_input([line | rest], row, acc) do
    acc =
      String.to_charlist(line)
      |> Enum.with_index()
      |> Enum.filter(fn {c, _i} -> c == ?# end)
      |> Enum.map(fn {_, col} -> {row, col} end)
      |> Enum.into(acc)

    parse_input(rest, row + 1, acc)
  end

  defp print_points(points) do
    {top, bottom, left, right} = bounding_rectangle(points)

    for row <- -top..bottom do
      IO.puts(
        :stderr,
        Enum.map(left..right, fn col ->
          if MapSet.member?(points, {row, col}), do: ?#, else: ?.
        end)
      )
    end
  end

  def main() do
    unless function_exported?(Runner, :main, 1), do: Code.compile_file("../runner.ex", __DIR__)
    success = Runner.main(Day23, System.argv())
    exit({:shutdown, if(success, do: 0, else: 1)})
  end
end

if Path.absname(:escript.script_name()) == Path.absname(__ENV__.file), do: Day23.main()

File: 2023/day23/Day23.jl
#!/usr/bin/env julia
# Copyright 2023 Google LLC
#
# Use of this source code is governed by an MIT-style
# license that can be found in the LICENSE file or at
# https://opensource.org/licenses/MIT.

"""# Advent of Code 2023 day 23
[Read the puzzle](https://adventofcode.com/2023/day/23)

Input is a grid of open spaces (`.`), walls (`#`), and downslopes (`<>^v`).  Start point is the
one open spot in the top row (one to the right of the edge) and destination is the one open spot
in the bottom row (one to the left of the edge).  Output is the longest path from start to
destination.  In part 1, when you're on a downslope you can only move in that direction.
In part 2, you can move in any free direction even when on a downslope.
"""
module Day23

function part1(lines)
  grid = parseinput(lines)
  start, target = CartesianIndex(1, 2), CartesianIndex(size(grid, 1), size(grid, 2) - 1)
  len, reached = longestpath(grid, start + DOWN, target, Set([start]), true)
  !reached && error("could not reach $target from $start")
  len
end

function part2(lines)
  grid = parseinput(lines)
  start, target = CartesianIndex(1, 2), CartesianIndex(size(grid, 1), size(grid, 2) - 1)
  longestpath2(grid, start, target)
end

function parseinput(lines)
  permutedims(reduce(hcat, collect.(lines)), (2, 1))
end

begin
  const LEFT = CartesianIndex(0, -1)
  const RIGHT = CartesianIndex(0, +1)
  const UP = CartesianIndex(-1, 0)
  const DOWN = CartesianIndex(+1, 0)
  const DIRS = (DOWN, RIGHT, UP, LEFT)
  const DOWNHILL = Dict('v' => DOWN, '>' => RIGHT, '^' => UP, '<' => LEFT)
end

function longestpath(grid, start, target, visited, downslopeonly)
  if start == target
    return 1, true
  end
  if !checkbounds(Bool, grid, start)
    return 0, false
  end
  cur = grid[start]
  if cur == '#'
    return 0, false
  end
  vis = visited ∪ (start,)
  dirs = if downslopeonly && haskey(DOWNHILL, cur)
    (DOWNHILL[cur],)
  else
    DIRS
  end
  lengths = Int[]
  for d in dirs
    next = start + d
    if next ∉ vis
      len, reached = longestpath(grid, next, target, vis, downslopeonly)
      if reached
        push!(lengths, 1+len)
      end
    end
  end
  isempty(lengths) ? (0, false) : (maximum(lengths), true)
end

function keypoints(grid)
  filter(x -> grid[x] != '#' && length(neighbors(grid, x)) > 2, eachindex(IndexCartesian(), grid))
end

function longestpath2(grid, start, target)
  keyp = vcat([start, target], keypoints(grid))
  nearby = Dict(x => Dict{CartesianIndex{2}, Int}() for x in keyp)
  for k in keyp
    for m in neighbors(grid, k)
      len = 1
      prev = k
      cur = m
      while cur ∉ keyp
        len += 1
        next = filter(x -> x != prev && x != k, neighbors(grid, cur))
        if length(next) != 1
          @show (k, m, cur, prev, next)
        end
        prev = cur
        cur = only(next)
      end
      nearby[k][cur] = len
    end
  end
  longest(nearby, start, target, CartesianIndex{2}[])
end

function longest(nearby, start, target, seen)
  if start == target
    return 0
  end
  lengths = Int[]
  seen = vcat(seen, start)
  for (k, v) in nearby[start]
    if k ∉ seen
      push!(lengths, v + longest(nearby, k, target, seen))
    end
  end
  return maximum(lengths; init=typemin(Int))
end

function neighbors(grid::Matrix{Char}, cur::CartesianIndex{2})
  filter(x -> checkbounds(Bool, grid, x) && grid[x] != '#', map(d -> cur + d, DIRS))
end

include("../Runner.jl")
@run_if_main
end
